cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java

```diff
@@ -155,19 +154,19 @@ public synchronized void create(final NamespaceMeta metadata) throws Exception {
       throw new NamespaceAlreadyExistsException(namespace);
     }
 
-    // if this namespace has custom mapping then validate the given custom mapping
+    // If this namespace has custom mapping then validate the given custom mapping
     if (hasCustomMapping(metadata)) {
       validateCustomMapping(metadata);
     }
 
-    // Namespace can be created. Check if the user is authorized now.
+    // Namespace can be created. Grant all the permissions to the user.
     Principal principal = authenticationContext.getPrincipal();
     privilegesManager.grant(namespace, principal, EnumSet.allOf(Action.class));
+
     // Also grant the user who will execute programs in this namespace all privileges on the namespace
     String executionUserName;
-    if (SecurityUtil.isKerberosEnabled(cConf)) {
-      ImpersonationInfo impersonationInfo = new ImpersonationInfo(metadata, cConf);
-      String namespacePrincipal = impersonationInfo.getPrincipal();
+    if (SecurityUtil.isKerberosEnabled(cConf) && NamespaceId.SYSTEM.equals(namespace)) {
+      String namespacePrincipal = metadata.getConfig().getPrincipal();
```

Comment:
but with your changes to ImpersonationInfo, it also does not do that any longer. I think you need to handle the case here.